### PR TITLE
Document the experimental registry signature endpoint

### PR DESCRIPTION
**Pull request by mfojtik**
This is WIP, I would like to add example of how you can sign the image using `atomic sign` or `atomic push` (@aweiteka to the rescue :-)

@openshift/team-documentation @miminar @legionus PTAL

Also i'm not sure what is the "right" section where this documentation should live (do we have something like "advanced registry" chapter?

The relevant Origin PR: https://github.com/openshift/origin/pull/12504

Also this documentation is for 3.4.1+ (do we have any label that says what is the target release? ;-)

### CODE REVIEW

**Review by aweiteka — CHANGES_REQUESTED** [review]
- Since this is a reference for someone integrating with the API it would be good to have a pointer to this page from the rest_api/index.adoc overview page. That would also tie in the story that these are useful OCP extensions found all in one place.
- Consider swapping the order so writing comes before reading to make for an example that actually works.

**Comment by mfojtik** [comment]
@aweiteka thanks! comments addressed.

File: install_config/registry/accessing_registry.adoc [hunk 1]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
 Digest: sha256:3662dd821983bc4326bee12caec61367e7fb6f6a3ee547cbaff98f77403cab55
 ----
 ====
+
+[[access-reading-and-writing-image-signatures]]
+=== Reading and Writing Image Signatures
+The {product-title} integrated Docker Registry provides an "extensions" endpoint that
+allows you to write and read image signatures. The image signatures are stored in
+{product-title} API, but this endpoint allows you to read and write them without
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: *API* - can we use etcd, storage or database instead?

File: install_config/registry/accessing_registry.adoc [hunk 2]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+information about the docker registry API.
+====
+
+==== Writing Image Signatures
+In order to add a new signature to the Image, users can use the HTTP `PUT` method together
+with the JSON payload sent the extensions endpoint:
+
+====
+----
+PUT /extensions/v2/{namespace}/{name}/signatures/sha256:{digest}
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: `digest` includes `sha256:` prefix

File: install_config/registry/accessing_registry.adoc [hunk 3]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+  "version": 2,
+  "type":    "atomic",
+  "name":    "sha256:4028782c08eae4a8c9a28bf661c0a8d1c2fc8e19dbaae2b018b21011197e1484@cddeb7006d914716e2728000746a0b23",
+  "content": "<base64 encoded signature>",
+}
+----
+====
+
+The `name` field contains the name of the image signature, which has to be unique and in
+the following format: `sha256:<digest>@<name>`. The `name` has to be 32 character long.
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: Explain that the first part is image name and the second is signature name. And that the first part must match the `<digest>` parameter of the endpoint.

File: install_config/registry/accessing_registry.adoc [hunk 4]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+In order to attach the signature to the Docker Image the user needs to have the `system:image-signer` role.
+====
+
+==== Reading Image Signatures
+Assuming that you have already pushed signed Docker Image into {product-title} integrated Docker
+Registry, you can read the signatures using the following command:
+
+====
+----
+GET /extensions/v2/{namespace}/{name}/signatures/sha256:{digest}
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: `digest` includes `sha256:`
https://github.com/docker/distribution/blob/master/docs/spec/api.md#content-digests

File: install_config/registry/accessing_registry.adoc [hunk 5]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+    "name":    "sha256:4028782c08eae4a8c9a28bf661c0a8d1c2fc8e19dbaae2b018b21011197e1484@cddeb7006d914716e2728000746a0b23",
+    "content": "<base64 encoded signature>",
+  }
+  ]
+}
+----
+====
+
+The `name` field contains the name of the image signature, which is unique and in the
+following format: `sha256:<digest>@<name>`. The `name` has to be 32 character long.
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: `<image-name>@<signature-name>` or something like that

**Review by miminar — CHANGES_REQUESTED** [review]
Just a few remarks.

File: install_config/registry/accessing_registry.adoc (hunk 1 above)
Comment (mfojtik) [inline install_config/registry/accessing_registry.adoc]: we can but we don't at this point? (don't want to make it more confusing)

Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: Let me explain. When you say API, I imagine just an interface, not a storage. So I find it hard to imagine that API stores something.So I wonder if there's any better word we can use as a replacement for the API. Another attempt: *key value store*?

Comment (aweiteka) [inline install_config/registry/accessing_registry.adoc]: How about, "The image signatures are stored in the {product-title} key-value store via API"?

File: install_config/registry/accessing_registry.adoc [hunk 6]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+=== Reading and Writing Image Signatures
+The {product-title} integrated Docker Registry provides an "extensions" endpoint that
+allows you to write and read image signatures. The image signatures are stored in
+{product-title} API, but this endpoint allows you to read and write them without
+having direct access to the {product-title} API endpoint.
+
+[NOTE]
+====
+This endpoint is experimental and not supported by the upstream Docker Registry project.
+See https://docs.docker.com/registry/spec/api/[upstream API documentation] for general
```
Comment (ahardin-rh) [inline install_config/registry/accessing_registry.adoc]: link:https://docs.docker.com/registry/spec/api/[upstream API documentation]

File: install_config/registry/accessing_registry.adoc [hunk 7]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+information about the docker registry API.
+====
+
```
Comment (ahardin-rh) [inline install_config/registry/accessing_registry.adoc]: Place this ID above the heading:
[[writing-image-signatures]]

File: install_config/registry/accessing_registry.adoc (hunk 3 above)
Comment (ahardin-rh) [inline install_config/registry/accessing_registry.adoc]: s/character/characters

File: install_config/registry/accessing_registry.adoc [hunk 8]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+}
+----
+====
+
+The `name` field contains the name of the image signature, which has to be unique and in
+the following format: `sha256:<digest>@<name>`. The `name` has to be 32 character long.
+
+[NOTE]
+====
+In order to attach the signature to the Docker Image the user needs to have the `system:image-signer` role.
```
Comment (ahardin-rh) [inline install_config/registry/accessing_registry.adoc]: s/Image/image,
s/needs to have/must have

File: install_config/registry/accessing_registry.adoc [hunk 9]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+====
+
+==== Reading Image Signatures
+Assuming that you have already pushed signed Docker Image into {product-title} integrated Docker
```
Comment (ahardin-rh) [inline install_config/registry/accessing_registry.adoc]: the signed Docker image into the {product-title} ...

File: install_config/registry/accessing_registry.adoc [hunk 10]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+====
+
+The `name` field contains the name of the image signature, which has to be unique and in
+the following format: `sha256:<digest>@<name>`. The `name` has to be 32 character long.
+
+[NOTE]
+====
+In order to attach the signature to the Docker Image the user needs to have the `system:image-signer` role.
+====
+
```
Comment (ahardin-rh) [inline install_config/registry/accessing_registry.adoc]: add this ID before the heading:
[[reading-image-signatures]]

File: install_config/registry/accessing_registry.adoc [hunk 11]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+====
+----
+$ curl http://<user>:<token>@<registry-endpoint>:5000/extensions/v2/<namespace>/<name>/signatures/sha256:<digest>
+----
+====
+
+Where the `namespace` represents the {product-title} project name or Docker Registry repository name and the `name`
+refers to the name of the Image. The `digest` represents the SHA256 checksum of the Image.
```
Comment (ahardin-rh) [inline install_config/registry/accessing_registry.adoc]: s/Image/image  (a few instances across the page)

File: install_config/registry/accessing_registry.adoc [hunk 12]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+{
+  "version": 2,
+  "type":    "atomic",
+  "name":    "sha256:4028782c08eae4a8c9a28bf661c0a8d1c2fc8e19dbaae2b018b21011197e1484@cddeb7006d914716e2728000746a0b23",
+  "content": "<base64 encoded signature>",
+}
+----
+====
+
+The `name` field contains the name of the image signature, which has to be unique and in
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: The `name` field can be confused with the `<name>` from the curl command above. Maybe `<isname> could be used instead there?

File: install_config/registry/accessing_registry.adoc [hunk 13]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+----
+====
+
+====
+----
+$ curl http://<user>:<token>@<registry-endpoint>:5000/extensions/v2/<namespace>/<name>/signatures/sha256:<digest>
+----
+====
+
+Where the `namespace` represents the {product-title} project name or Docker Registry repository name and the `name`
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: `name` refers to the repository, not the image

File: install_config/registry/accessing_registry.adoc (hunk 1 above)
Comment (mfojtik) [inline install_config/registry/accessing_registry.adoc]: fixed.

File: install_config/registry/accessing_registry.adoc (hunk 12 above)
Comment (mfojtik) [inline install_config/registry/accessing_registry.adoc]: it says "name field"

File: install_config/registry/accessing_registry.adoc (hunk 11 above)
Comment (mfojtik) [inline install_config/registry/accessing_registry.adoc]: tried to fix them all.. in some cases I was refering to image and in some cases to API Image object... I dunno if we make distincion between them.

File: install_config/registry/accessing_registry.adoc [hunk 14]
```diff
@@ -272,3 +272,95 @@ cf2616975b4a: Image successfully pushed
...
+[NOTE]
+====
+This endpoint is experimental and not supported by the upstream Docker Registry project.
+See https://docs.docker.com/registry/spec/api/[upstream API documentation] for general
+information about the docker registry API.
+====
+
+==== Writing Image Signatures
+In order to add a new signature to the Image, users can use the HTTP `PUT` method together
+with the JSON payload sent the extensions endpoint:
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: The sentence makes sense to me until **`sent`**.

**Comment by mfojtik** [comment]
@ahardin-rh @miminar tried to address all comments, PTAL

File: install_config/registry/accessing_registry.adoc [hunk 15]
```diff
@@ -272,3 +272,99 @@ cf2616975b4a: Image successfully pushed
...
+====
+This endpoint is experimental and not supported by the upstream Docker Registry project.
+See link:https://docs.docker.com/registry/spec/api/[upstream API documentation] for
+general information about the Docker Registry API.
+====
+
+[[writing-image-signatures]]
+==== Writing Image Signatures
+In order to add a new signature to the Image, users can use the HTTP `PUT` method together
+with the JSON payload sent the extensions endpoint:
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: I understand this sentence until **`sent`**.

File: install_config/registry/accessing_registry.adoc [hunk 16]
```diff
@@ -272,3 +272,99 @@ cf2616975b4a: Image successfully pushed
...
+
+====
+----
+PUT /extensions/v2/{namespace}/{name}/signatures/{digest}
+----
+====
+
+====
+----
+$ curl -X PUT --data @signature.json http://<user>:<token>@<registry-endpoint>:5000/extensions/v2/<namespace>/<name>/signatures/sha256:<digest>
```
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: To be consistent, replace all the occurences of `sha256:<digest>` with just `<digest>`.

**Review by miminar — CHANGES_REQUESTED** [review]
Just a few minor glitches.

**Comment by adellape** [comment]
[Document: Add registry endpoint for reading and writing image signatures](https://trello.com/c/RacGhyYi/445-document-add-registry-endpoint-for-reading-and-writing-image-signatures)

File: install_config/registry/accessing_registry.adoc (hunk 15 above)
Comment (miminar) [inline install_config/registry/accessing_registry.adoc]: Still valid.

**Comment by adellape** [comment]
Closing this and continuing via https://github.com/openshift/openshift-docs/pull/4623, PTAL there.

**Pull request closed without merging**